### ISSUE TITLE
Spacing between outline and text

### ISSUE THREAD

**Issue by Sn0wCrack** [libass/libass#289]
I'm noticing using \bord on anything higher than 1 or 2 seems to give a gap between the text itself and the outline.

I've currently tested against MPC-HC (built in VSFilter, XySubFilter and AssFilterMod), mpv and VLC, this issue only seems to appear on the libass based renderers.

I've also tried with ScaledBorderAndShadow on yes and no and neither option seems to affect anything.

Screenshots:
libass:
![](https://i.imgur.com/D8MqhW8.png)

VSFilter:
![](https://i.imgur.com/34z3Bnf.png)

Subtitle Lines:
```
Style: Show,Arial,18,&H00FFFFFF,&H000000FF,&H00000000,&H00000000,0,0,0,0,100,100,0,0,1,0,0,2,20,20,20,1

Dialogue: 0,0:00:50.32,0:00:52.02,Show,Title,0,0,0,,{\an4\bord22\fs66\fsp1\fnQuicksand Bold\blur6\c&H52F6FB&\3c&H52F6FB&\pos(505,229.333)}Please tell me!
Dialogue: 1,0:00:50.32,0:00:52.02,Show,Title,0,0,0,,{\an4\shad5\bord7\fs66\fsp1\fnQuicksand Bold\blur0.7\c&H9A4975&\3c&H9A4975\4c&H060A1B&\pos(505,229.333)}Please tell me!
Dialogue: 0,0:00:50.32,0:00:52.02,Show,Title,0,0,0,,{\an4\bord27\fs110\blur6\c&H52F6FB&\3c&H52F6FB&\pos(505,882)}■ ■ ■ ■ ■ ■ ■ ■ ■ ■ ■ ■
Dialogue: 1,0:00:50.32,0:00:52.02,Show,Title,0,0,0,,{\an4\shad5\bord8\fs110\blur0.7\c&HA052E1&\3c&A052E1\4c&H060A1B&\pos(505,882)}■ ■ ■ ■ ■ ■ ■ ■ ■ ■ ■ ■
```

**Comment by astiob** [comment]
For some more fun, try this in VSFilter:

```
Dialogue: 0,0:00:00.00,0:00:10.00,Default,,0,0,0,,{\an7\pos(0,0)\c&HFFFFFF&\p1}m 0 0 l 1920 0 1920 1080 0 1080
Dialogue: 0,0:00:00.00,0:00:10.00,Default,,0,0,0,,{\fs60\bord5\c&H000000&\2a&H01&\3c&H000000&}test
```

(The first line just creates a white background so that the problem is more easily seen.)

![gap](https://user-images.githubusercontent.com/515193/32356875-c8b262c4-c040-11e7-8c79-3b84d6919d20.png)

The fundamental cause of this effect is the way we (and VSFilter) output images: they’re all layered on top of each other, and in particular text proper is layered on top of text border. At the edges, text has semitransparent pixels, and when they overlap with border pixels, all sorts of weird things happen. The particular darkening you are seeing may be caused by a naïve countermeasure that’s currently in place in libass, which was introduced to fix some other (probably more common) cases but obviously breaks this case instead.

(This is also discussed a little bit in #76.)

Back in (the monstrous) [Google Code pull request 78](http://code.google.com/archive/p/libass/issues/78), I wrote a patch that was meant to fix this once and for all, without any downsides. Unfortunately, it turned out that I didn’t take into account karaoke effects, which are applied at a much later stage in the libass rendering pipeline, so, while my patch did fix all cases where karaoke wasn’t used, it introduced very visible distortions into karaoke when there was a high contrast between the karaoke colour and the primary text colour. Later, while discussing #76, I also noticed that my patch wasn’t mathematically sound when blur was involved, but I promptly fixed this. It still remains to fix karaoke effects, and then we’ll be able to finally put this issue to rest for good.

By the way, your sample code does not exactly match the sample images: the code is missing the “gyaruko-chan” text but has the “Please tell me!” text, whereas the images show the opposite. It doesn’t really matter though, as the problem is clearly seen in the squares behind “gyaruko-chan”.

**Comment by rcombs** [comment]
Fixed in 12cf524b831289cf67ba3432264fa01a69c3bbb5.
